### PR TITLE
Make Suggest edits button link to core repo docstring source

### PR DESCRIPTION
**Pull request by snehankekre**
This PR is a very early WIP to make the Suggest edits button on API reference pages link to the original source of the docstring in the core Streamlit repo.

[#71f8d97](https://github.com/streamlit/docs/commit/71f8d9789c53a3632409a2a85c0cb206f001f850) creates an **additional** "Suggest edits" button at the bottom of API reference pages. This is not the desired behavior. Ideally, for API reference pages, we need to override the Helpful component and pass it the GitHub URL instead of `sourcefile`.

@imjuangarcia The Python part of this PR is complete. I could use your input on the JS side of this PR. We need to pass the `functionObject.source` object (containing a Streamlit GitHub URL) from `autofunction.js` to either the Helpful or SuggestEdits component. This should be done only for API reference pages containing docstrings. i.e. pages whose Markdown filename ends with `/library/api/*/*.md`, where * is an alphanumeric string; or equivalently, pages that use the Autofunction component.

### CODE REVIEW

**Comment by imjuangarcia** [comment]
@snehankekre What do you think of the approach taken on fef2f29? I'm rendering a new `<Helpful />` utility specific to the `<Autofunction>` component and, when that one is present on the page, I remove the general one we have for all `[...slug]` pages.

**Comment by snehankekre** [comment]
@imjuangarcia This is exactly what I in mind! Do you think we're ready to merge?

**Comment by snehankekre** [comment]
@imjuangarcia Noticed a bug in pages that include content below the `<Autofunction />` component. The Helpful utility is displayed (mid-page) before the rest of the content. See [1](https://deploy-preview-223--streamlit-docs.netlify.app/library/api-reference/widgets/st.button), [2](https://deploy-preview-223--streamlit-docs.netlify.app/library/api-reference/widgets/st.radio), [3](https://deploy-preview-223--streamlit-docs.netlify.app/library/api-reference/widgets/st.slider):

image

**Comment by imjuangarcia** [comment]
> @imjuangarcia Noticed a bug in pages that include content below the `<Autofunction />` component. The Helpful utility is displayed (mid-page) before the rest of the content

@snehankekre Great catch, thank you! The approach I was thinking of won't work, then. I fixed it on 7e41a12 using React Context to store the Autofunction source file globally. I'm looping @tvst in here so he can review the JS piece!

File: python/generate.py [hunk 1]
```diff
@@ -95,6 +95,19 @@ def get_function_docstring_dict(func, funcname, signature_prefix):
                 return_obj['description'] = parse_rst(returns.description) if returns.description else ''
                 return_obj['return_name'] = returns.return_name
                 description['returns'].append(return_obj)
+        
+        description['source'] = []
+        repo_prefix = "https://github.com/streamlit/streamlit/blob/develop/lib"
```
Comment (tvst) [inline python/generate.py:100]: @snehankekre : Move into top-level function.

File: python/generate.py [hunk 2]
```diff
@@ -95,6 +95,19 @@ def get_function_docstring_dict(func, funcname, signature_prefix):
...
+
+        try: # For Streamlit commands wrapped by decorator
+            path_parts = func.__wrapped__.__code__.co_filename.partition("/streamlit")
+            line_number = func.__wrapped__.__code__.co_firstlineno
```
Comment (tvst) [inline python/generate.py:104]: @snehankekre : May want to do a `while` loop checking whether there's a `.__wrapped__` . In fact, you could even remove the try/except this way:

```
while '__wrapped__' in func:
  func = func.__wrapped__

path_parts = func.__code__.co_filename.partition("/streamlit")
line_number = func.__code__.co_firstlineno
```

File: python/generate.py [hunk 3]
```diff
@@ -95,6 +95,19 @@ def get_function_docstring_dict(func, funcname, signature_prefix):
...
+        except:
+            path_parts = func.__code__.co_filename.partition("/streamlit")
+            line_number = func.__code__.co_firstlineno
+        
+        description['source'] = repo_prefix + path_parts[1] + path_parts[2]+ f'#L{line_number}'
```
Comment (tvst) [inline python/generate.py:109]: @snehankekre 

Use f-string:
```
description['source'] = f'{repo_prefix}{path_parts[1]}{path_parts[2]}#L{line_number}'
```

or join:
```
description['source'] = ''.join([
  repo_prefix,
  path_parts[1],
  path_parts[2],
  f'#L{line_number}',
])
```

File: pages/[...slug].js [hunk 4]
```diff
@@ -72,11 +71,14 @@ export default function Article({
   let versionWarning;
   let currentLink;
   let sourceFile;
+  const autoFunctionSourceFile = useAppContext();
```
Comment (tvst) [inline pages/[...slug].js:74]: This could be cleaner:

```
const { sourceFile } = ...
```

**Review by tvst — APPROVED** [review]
LGTM after changes.

File: pages/[...slug].js (hunk 4 above)
Comment (imjuangarcia) [reply pages/[...slug].js]: Fixed on 709f4a9

File: python/generate.py (hunk 1 above)
Comment (snehankekre) [reply python/generate.py]: Fixed on f73e87b.

File: python/generate.py (hunk 2 above)
Comment (snehankekre) [reply python/generate.py]: So much cleaner, thank you! I did:

```
while '__wrapped__' in func.__dict__:
  func = func.__wrapped__

path_parts = func.__code__.co_filename.partition("/streamlit")
line = func.__code__.co_firstlineno
```

File: python/generate.py (hunk 3 above)
Comment (snehankekre) [reply python/generate.py]: I used join on f73e87b.

**Pull request merged**